src/ray/object_manager/object_directory.cc

```diff
@@ -8,34 +8,43 @@ ObjectDirectory::ObjectDirectory(boost::asio::io_service &io_service,
 
 namespace {
 
-std::vector<ClientID> UpdateObjectLocations(
-    std::unordered_set<ClientID> &client_ids,
-    const std::vector<ObjectTableDataT> &location_history,
-    const ray::gcs::ClientTable &client_table) {
+/// Process a suffix of the object table log and store the result in
+/// client_ids. This assumes that client_ids already contains the result of the
+/// object table log up to but not including this suffix. This also stores a
+/// bool in has_been_created indicating whether the object has ever been
+/// created before.
+void UpdateObjectLocations(const std::vector<ObjectTableDataT> &location_history,
+                           const ray::gcs::ClientTable &client_table,
+                           std::unordered_set<ClientID> *client_ids,
+                           bool *has_been_created) {
```

Comment:
This unfortunately doesn't seem to play too nicely with https://github.com/ray-project/ray/pull/3499 because sometimes we evict the keys so they will appear to have never been created. cc @ericl

Reply:
Hmm that's true...I can't really think of a foolproof way around that except to fail the object after some number of attempts.